### PR TITLE
[BEAM-13152] [BEAM-13163] [Playground] Run Go code

### PR DESCRIPTION
**Pull request by pavel-avilov**
Add Go config;
Rebase the file system;
Support go sdk;

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/cmd/server/controller.go [hunk 1]
```diff
@@ -208,7 +208,7 @@ func setupCompileBuilder(lc *fs_tool.LifeCycle, sdk pb.Sdk, executorConfig *envi
 		WithCompiler()
 
 	switch sdk {
-	case pb.Sdk_SDK_JAVA:
+	case pb.Sdk_SDK_JAVA, pb.Sdk_SDK_GO:
```
Comment (AydarZaynutdinov) [inline playground/backend/cmd/server/controller.go:211]: we could remove the `switch` statement. These actions should be done for each SDK, isn't it?

File: playground/backend/internal/fs_tool/go_fs_test.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		pipelineId uuid.UUID
+		workingDir string
+	}
+	tests := []struct {
+		name string
+		args args
+		want *LifeCycle
+	}{
+		{
+			name: "newGoLifeCycle",
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/fs_tool/go_fs_test.go:42]: Need to add comments for each test case.

File: playground/backend/internal/fs_tool/java_fs_test.go [hunk 3]
```diff
@@ -103,7 +103,7 @@ func Test_executableName(t *testing.T) {
 		{
 			name: "all success",
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/fs_tool/java_fs_test.go:104]: Rename it to `get executable name`

File: playground/backend/internal/environment/environment_service.go [hunk 4]
```diff
@@ -179,8 +179,6 @@ func createExecutorConfig(apacheBeamSdk pb.Sdk, configPath string) (*ExecutorCon
 			getEnv(SLF4jKey, defaultSLF4j),
 		}, ":")
 		executorConfig.RunArgs[1] += jars
-	case pb.Sdk_SDK_GO:
```
Comment (daria-malkova) [inline playground/backend/internal/environment/environment_service.go:182]: if you removed it from here, how executorConfig will be created?

File: playground/backend/internal/fs_tool/compiled_tool.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/compiled_tool.go:1]: not sure about the name of the file. what is the compiled tool?

File: playground/backend/internal/fs_tool/compiled_tool.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+const (
+	baseFileFolder     = "executable_files"
+	sourceFolderName   = "src"
+	compiledFolderName = "bin"
+)
+
+// newCompiledLifeCycle creates LifeCycle with compiled SDK environment.
+func newCompiledLifeCycle(pipelineId uuid.UUID, workingDir string, executableFileExtension string, compiledFileExtension string) *LifeCycle {
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/compiled_tool.go:30]: compiled life cycle also doesn't sound good
javaLifeCycle, goLifeCycle sounds fine, compiled is not clear what it is

File: playground/backend/internal/fs_tool/fs.go [hunk 7]
```diff
@@ -109,6 +111,14 @@ func (l *LifeCycle) GetAbsoluteExecutableFilePath() string {
 	return absoluteFilePath
 }
 
+// GetAbsoluteCompiledFilePath returns absolute filepath to compiled file (/path/to/workingDir/executable_files/{pipelineId}/bin/{pipelineId}.class for java SDK).
+func (l *LifeCycle) GetAbsoluteCompiledFilePath() string {
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/fs.go:115]: what is the "compiled" here? I assume that here it is a compiled go file, but it is not clear

File: playground/backend/internal/fs_tool/fs.go [hunk 8]
```diff
@@ -117,5 +127,9 @@ func (l *LifeCycle) GetAbsoluteExecutableFilesFolderPath() string {
 
 // getFileName returns fileName by pipelineId and fileType ({pipelineId}.java for java SDK).
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/fs.go:128]: change the description, go is different

File: playground/backend/internal/fs_tool/fs.go [hunk 9]
```diff
@@ -117,5 +127,9 @@ func (l *LifeCycle) GetAbsoluteExecutableFilesFolderPath() string {
...
 func getFileName(pipelineId uuid.UUID, fileType string) string {
-	return fmt.Sprintf("%s.%s", pipelineId, fileType)
+	if fileType != "" {
+		return fmt.Sprintf("%s.%s", pipelineId, fileType)
+	} else {
+		return pipelineId.String()
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/fs.go:133]: not very clear why there will be no extension here

File: playground/backend/internal/fs_tool/fs.go [hunk 10]
```diff
@@ -117,5 +127,9 @@ func (l *LifeCycle) GetAbsoluteExecutableFilesFolderPath() string {
 
 // getFileName returns fileName by pipelineId and fileType ({pipelineId}.java for java SDK).
 func getFileName(pipelineId uuid.UUID, fileType string) string {
-	return fmt.Sprintf("%s.%s", pipelineId, fileType)
+	if fileType != "" {
+		return fmt.Sprintf("%s.%s", pipelineId, fileType)
+	} else {
```
Comment (daria-malkova) [inline playground/backend/internal/fs_tool/fs.go:132]: and in general I would use inheritance (we have goLifeCycle and javaLC) to override method and wouldn't use if-else

File: playground/backend/internal/validators/go_validators.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validators
+
+// GetGoValidators return validators methods that should be applied to Go code
+func GetGoValidators() *[]Validator {
+	//TODO: If there is an idea for go validators in the future, add them here
```
Comment (daria-malkova) [inline playground/backend/internal/validators/go_validators.go:20]: You know it is not the best idea to leave to-do's
Write some validator, for example check go fmt

**Review by daria-malkova — CHANGES_REQUESTED** [review]

File: playground/backend/internal/fs_tool/fs.go (hunk 9 above)
Comment (pavel-avilov) [reply playground/backend/internal/fs_tool/fs.go:133]: cause compiled go files have no extension and without this condition, the name of the binary file will be like "name."

File: playground/backend/internal/fs_tool/fs.go (hunk 10 above)
Comment (pavel-avilov) [reply playground/backend/internal/fs_tool/fs.go:132]: got it

File: playground/backend/internal/environment/environment_service.go (hunk 4 above)
Comment (pavel-avilov) [reply playground/backend/internal/environment/environment_service.go:182]: The config is initialized on line 169. In this switch case, the necessary arguments are added to the default config. The executorConfig for GO doesn't need these additional arguments, so I remove this case.

Comment (daria-malkova) [reply playground/backend/internal/environment/environment_service.go:182]: can be questions if just to delete the case with Go. Make it clear, return executorConfig

**Review by daria-malkova — CHANGES_REQUESTED** [review]

File: playground/backend/internal/environment/environment_service.go (hunk 4 above)
Comment (KhaninArtur) [reply playground/backend/internal/environment/environment_service.go:182]: We can just leave this case empty and add a comment that Go doesn't need any additional arguments from the config file

File: playground/backend/internal/fs_tool/compiling_tool.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (KhaninArtur) [inline playground/backend/internal/fs_tool/compiling_tool.go:1]: The filename looks a bit wrong among others in the same folder. Can we change it with something more appropriate like `compiling_lc.go`? Maybe it is not the best option, let's think about it.

File: playground/backend/internal/fs_tool/go_fs_test.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	tests := []struct {
+		name string
+		args args
+		want *LifeCycle
+	}{
+		{
+			// Test case with calling newGoLifeCycle method with correct pipelineId and workingDir.
+			// As a result, want to receive an expected go life cycle.
+			name: "newGoLifeCycle",
```
Comment (KhaninArtur) [inline playground/backend/internal/fs_tool/go_fs_test.go:44]: Do we have a negative test case here?

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

File: playground/backend/internal/fs_tool/go_fs_test.go (hunk 13 above)
Comment (pavel-avilov) [reply playground/backend/internal/fs_tool/go_fs_test.go:44]: This is a constructor and I think we cannot write a negative test case here

File: playground/backend/internal/fs_tool/compiling_tool.go (hunk 12 above)
Comment (pavel-avilov) [reply playground/backend/internal/fs_tool/compiling_tool.go]: Done.

File: playground/backend/internal/environment/environment_service.go (hunk 4 above)
Comment (pavel-avilov) [reply playground/backend/internal/environment/environment_service.go:182]: Done

**Review by KhaninArtur — APPROVED** [review]
LGTM

**Review by daria-malkova — APPROVED** [review]
LGTM

**Review by ilya-kozyrev — APPROVED** [review]

**Pull request closed without merging**